libeos-payg/manager.c

```diff
@@ -822,12 +850,45 @@ epg_manager_clear_code (EpgProvider  *provider,
 
   clear_expiry_timer (self);
 
-  /* Kick off an asynchronous save. */
-  epg_manager_save_state_async (provider, self->cancellable, NULL, NULL);
+  /* Kick off an asynchronous save.
+   *
+   * FIXME: pass self->cancellable; see comment in
+   * epg_manager_shutdown_async().
+   */
+  g_assert_cmpuint (self->pending_internal_save_state_calls, <, G_MAXUINT64);
+  self->pending_internal_save_state_calls++;
+  epg_manager_save_state_async (provider, NULL,
+                                internal_save_state_cb, NULL);
 
   return TRUE;
 }
 
+static void
+internal_save_state_cb (GObject      *source_object,
+                        GAsyncResult *result,
+                        gpointer      user_data)
+{
+  EpgManager *self = EPG_MANAGER (source_object);
+  EpgProvider *provider = EPG_PROVIDER (self);
+  g_autoptr(GError) local_error = NULL;
+
+  if (!epg_manager_save_state_finish (provider, result, &local_error))
+    g_warning ("save_state failed: %s", local_error->message);
+
+  g_return_if_fail (self->pending_internal_save_state_calls > 0);
+  self->pending_internal_save_state_calls--;
+
+  if (self->pending_internal_save_state_calls == 0 &&
```

Comment:
This block deserves a bit of a comment, I think, since it’s effectively kicking off another save operation directly after a previous one has (potentially successfully) finished.

Reply:
Hmm. Actually, this is only needed because if there was more than one pending call, it's not guaranteed that the later one "wins". Perhaps I should change all this logic to allow:

* at most 1 save operation in flight (with 1 or more tasks chained to it)
* a possibly-empty queue of requests to be serviced (ideally all at once) when the in-flight request completes

It seems like a lot of :keyboard: but it would have the advantage of being more robust in the face of unexpected power loss. I only have limited appetite for caring about this, though, because in ordinary use it's not possible to have >1 save at a time (since the UI only allows you to enter a single code, only when the machine is locked).

Reply:
The less laborious option is to make shutdown_async something like:

```
if no saves in flight:
    return True # nothing to do
else:
    wait for last save to finish, propagate its result
```

This changes the behaviour in the case where writing back the state would fail, but the state hasn't changed so there's no need to write it back. It also ignores the problem that if two saves are in flight it's not defined which would finish first. But it is a lot easier…

Reply:
I implemented:

```
if no saves in flight:
   save once more
else:
   wait for last to finish & propagate result
```